include/examples/quadruped_3dof_limb.h

```diff
@@ -0,0 +1,323 @@
+        }
+
+        /* Hip Angle */
+        float sin_th_ab, cos_th_ab;
+        sincosf(theta_abd, &sin_th_ab, &cos_th_ab);
+
+        // Move toe into hip plane for hip knee calculation
+        // Projected Leg Vector on the xz plane of the HIP
+        // [1,0,0;0,0,1] (Rx(th) (xToe - d0) - d1  - [0;L1;0])
+        // Projection Toe
+        // [1,0,0;0,0,1] Rx(th)
+        Eigen::Matrix<float, 2, 3> toeProjMatrix;
+        toeProjMatrix << 1, 0, 0,
+            0, -sin_th_ab, cos_th_ab; // angle sum identity 
+        Eigen::Vector2f projToe = toeProjMatrix * pos - config_.actuator_offsets[1]({0, 2});
+        // Hip-Knee Plane Leg Length
+        float r_sq = pos_sq_norm - L1 * L1;
+        float r = sqrt(r_sq);
+        // Compute & Clip Angle
+        float theta_hip = atan2(-projToe.x(), -projToe.y()) + acos(-(L3 * L3 - L2 * L2 - r_sq) / (2.0 * L2 * r));
+        theta_hip = quad_example_helpers::angle_mod(theta_hip);
+        
+        /* Return Computed Angles */
+        return {theta_abd, theta_hip, theta_knee};
+    }
+
+    // Slight optimization over individual calculations
+    void FKAndJacobianImpl(
+        std::vector<float> joint_positions) override
+    {
+        using quad_example_helpers::Skew;
+        Eigen::Matrix3f R_AB(Eigen::AngleAxisf(joint_positions[0], 
+                                               Eigen::Vector3f::UnitX()));
+        Eigen::Matrix3f R_BC(Eigen::AngleAxisf(joint_positions[1], 
+                                               Eigen::Vector3f::UnitY()));
+        Eigen::Matrix3f R_CD(Eigen::AngleAxisf(joint_positions[2], 
+                                               Eigen::Vector3f::UnitY()));
+
+        Eigen::Matrix3f abduc_frame = R_AB;
+        Eigen::Matrix3f hip_frame = abduc_frame * R_BC;
+        Eigen::Matrix3f knee_frame = hip_frame * R_CD;
+        
+        Eigen::Vector3f abduc_pos = config_.actuator_offsets[0];
+        Eigen::Vector3f hip_pos = abduc_pos + 
+                                  abduc_frame * config_.actuator_offsets[1];
+        Eigen::Vector3f knee_pos = hip_pos +
+                                   hip_frame * config_.actuator_offsets[2]; 
+
+        Eigen::Matrix3f xhat_skew = Skew(Eigen::Vector3f::UnitX());
+        Eigen::Matrix3f yhat_skew = Skew(Eigen::Vector3f::UnitY());
+
+        Eigen::Vector3f toe_position;
+        Eigen::Matrix3f toe_jacobian;
+        
+        toe_position = config_.actuator_offsets[3];
+        toe_jacobian.col(2) = knee_frame * yhat_skew * toe_position;
+        toe_position = config_.actuator_offsets[2] + R_CD * toe_position;
+        toe_jacobian.col(1) = hip_frame * yhat_skew * toe_position;
+        toe_position = config_.actuator_offsets[1] + R_BC * toe_position;
+        toe_jacobian.col(0) = abduc_frame * xhat_skew * toe_position;
+        toe_position = config_.actuator_offsets[0] + R_AB * toe_position;
+
+        // Set the cached kinematics
+        fk_.set(toe_position);
```

Comment:
Why are we expecting the user to set? Seems like bad design.

Reply:
Could use a `std::pair<EndEffectorOutput, Eigen::MatrixXf>` output, or it could be hidden behind the `=operator` overload, but figured that this function is meant to be more advanced anyway.
The example is there to guide them through it, but I know it isn't super smooth.

Reply:
I'd use the std::pair.

Reply:
Ok, assuming that is the case, two small questions exist.
first,medium: with my current implementation, I would need to also use std::optional i.e.
`
std::optional<std::pair<EndEffectorOutput, Eigen::MatrixXf>
`
or

`
std::pair<std::optional<EndEffectorOutput>, std::optional<Eigen::MatrixXf>>
`
second, minor : Can the optimizer remove that copy? can I ensure no copy?

Reply:
Whats the use case for returning `std::nullopt`?

For the minor question, here is a discussion. But the answer is the optimizer will likely remove the copy. https://stackoverflow.com/a/4643721

Reply:
The functionality as it currently in the `Update` function is to always run the FKAndJacobian first, and if the FK or Jac is not set, then run FKImpl and JacImpl. If `FKAndJacobianImpl` returns a pair, I would like to be able to tell that it ran the default implementation (which doesn't actually calculate).

Reply:
I think the `std::optional<std::pair<...>>` is the correct way to do it. Ideally there would be some way to programmatically check if the function was overridden but I don't think that is possible.